### PR TITLE
[V1][TPU] Integrate the new ragged paged attention kernel with vLLM v1 on TPU

### PR DESCRIPTION
**Pull request by vanbasten23**
This PR integrates the new ragged paged attention kernel with vLLM v1 on TPU. In particular, this PR

- Update torch_xla pin to the latest
- Update pallas.py in v1 to use the new ragged paged attention kernel instead of the 3 separate kernels in v0.
- Combine prompt and decode steps into one single step in tpu_model_runner.py, similar to what GPU does today.

Test plan:
- $ VLLM_USE_V1=1 python vllm/examples/offline_inference/basic.py 2>&1 | tee out.txt
- $ VLLM_USE_V1=1 pytest -s -v vllm/tests/entrypoints/llm/test_accuracy.py::test_lm_eval_accuracy_v1_engine 2>&1 | tee out.txt

### CODE REVIEW

File: vllm/v1/core/scheduler.py [hunk 1]
```diff
@@ -497,6 +497,7 @@ def update_from_output(
 
             req_index = model_runner_output.req_id_to_index[req_id]
             generated_token_ids = sampled_token_ids[req_index]
+            # print(f'xw32 update_from_output {req_index=}, {len(generated_token_ids)=}')
```
Comment (vanbasten23) [inline vllm/v1/core/scheduler.py:500]: remove

**Comment by vanbasten23** [comment]
@alexm-redhat @WoosukKwon , there are 2 issues currently. One is running `vllm/tests/entrypoints/llm/test_accuracy.py::test_lm_eval_accuracy_v1_engine` is very slow (2h, probably due to excessive compiling), which I'm investigating. Another issue is the test `vllm/tests/entrypoints/llm/test_accuracy.py::test_lm_eval_accuracy_v1_engine` actually fails. Any suggestion on how to find a smaller repro in order to debug?

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -730,190 +533,85 @@ def load_model(self) -> None:
                                    fullgraph=True,
                                    dynamic=False)
```
Comment (vanbasten23) [inline vllm/v1/worker/tpu_model_runner.py:534]: @WoosukKwon , do you remember the reason why we set dynamic=False?

**Comment by vanbasten23** [comment]
cc @bvrockwell

**Comment by vanbasten23** [comment]
@vllm-v-team@google.com

**Comment by vanbasten23** [comment]
hey @mgoin , it's my first PR in vLLM repo. I see "pre-commit / pre-commit (pull_request)" in the CI is red, it seems that it complains the format and mypy. For formatting, is there a linter I can use in vLLM?

**Comment by mgoin** [comment]
Hey @vanbasten23 please install precommit using these directions https://docs.vllm.ai/en/latest/contributing/overview.html#testing

```
pip install -r requirements-dev.txt
pre-commit install --hook-type pre-commit --hook-type commit-msg
```

Then on your next commit it will apply

File: requirements-tpu.txt [hunk 3]
```diff
@@ -17,9 +17,8 @@ ray[default]
...
-torch_xla[tpu, pallas] @ https://storage.googleapis.com/pytorch-xla-releases/wheels/tpuvm/torch_xla-2.7.0.dev20250124-cp39-cp39-linux_x86_64.whl ; python_version == "3.9"
-torch_xla[tpu, pallas] @ https://storage.googleapis.com/pytorch-xla-releases/wheels/tpuvm/torch_xla-2.7.0.dev20250124-cp310-cp310-linux_x86_64.whl ; python_version == "3.10"
-torch_xla[tpu, pallas] @ https://storage.googleapis.com/pytorch-xla-releases/wheels/tpuvm/torch_xla-2.7.0.dev20250124-cp311-cp311-linux_x86_64.whl ; python_version == "3.11"
+
+torch==2.7.0.dev20250220+cpu
+torch_xla[tpu, pallas] @ https://storage.googleapis.com/pytorch-xla-releases/wheels/tpuvm/torch_xla-2.7.0.dev20250220+cxx11-cp39-cp39-linux_x86_64.whl ; python_version == "3.9"
+torch_xla[tpu, pallas] @ https://storage.googleapis.com/pytorch-xla-releases/wheels/tpuvm/torch_xla-2.7.0.dev20250220+cxx11-cp310-cp310-linux_x86_64.whl ; python_version == "3.10"
+torch_xla[tpu, pallas] @ https://storage.googleapis.com/pytorch-xla-releases/wheels/tpuvm/torch_xla-2.7.0.dev20250220+cxx11-cp311-cp311-linux_x86_64.whl ; python_version == "3.11"
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply requirements-tpu.txt:24]: Yeah, the new kernel is in the latest torch_xla wheel. That's why I need to bump the torch_xla version.

**Review by mgoin — APPROVED** [review]
I am able to get good evaluations with Qwen/Qwen2-1.5B-Instruct on gsm8k with a small number of samples and with a small batch size. The attention kernel clearly has issues at the moment with compilation time/memory usage, but this interface is where we should go. So I think we should land this as-is and quickly iterate with new kernels to improve the usability and fix bugs, rather than stay in the emulated V0 style we have on main currently.

Thank you for your great work!

**Comment by vanbasten23** [comment]
Thanks @mgoin for reviewing the PR.

**Comment by vanbasten23** [comment]
Hi @mgoin , could you help merge the PR? I don't see a merge button on my side.

**Comment by mgoin** [comment]
Thanks for the ping, yes you need committer status to merge, which I'll handle. Let me quickly chat with @alexm-redhat before merging

File: vllm/v1/attention/backends/pallas.py [hunk 4]
```diff
@@ -4,13 +4,16 @@
...
+# Required to register custom ops.
+import torch_xla.experimental.custom_kernel  # noqa: F401
 
 from vllm.attention.backends.abstract import (AttentionBackend, AttentionImpl,
-                                              AttentionLayer,
-                                              AttentionMetadata, AttentionType)
+                                              AttentionLayer, AttentionType)
 from vllm.attention.backends.utils import CommonAttentionState
 
+NUM_QUERIES_PER_BLOCK = 16
```
Comment (alexm-redhat) [inline vllm/v1/attention/backends/pallas.py:14]: A quick doc about these numbers would be helpful

File: vllm/v1/attention/backends/pallas.py [hunk 5]
```diff
@@ -164,135 +127,47 @@ def forward(
...
+            query,
+            key_cache,
+            value_cache,
+            attn_metadata.context_lens,
+            attn_metadata.block_tables,
+            attn_metadata.query_start_loc,
+            attn_metadata.num_seqs,
+            num_kv_pages_per_block=NUM_KV_PAGES_PER_BLOCK,
+            num_queries_per_block=NUM_QUERIES_PER_BLOCK,
+            use_kernel=False,
```
Comment (alexm-redhat) [inline vllm/v1/attention/backends/pallas.py:167]: a quick doc about use_kernel would be helpful

**Review by alexm-redhat — APPROVED** [review]
@vanbasten23 thanks for the hard work on this, the PR looks good to land! Just some small comments, but you can address them also after landing.

File: vllm/v1/attention/backends/pallas.py (hunk 5 above)
Comment (vanbasten23) [reply vllm/v1/attention/backends/pallas.py:167]: resolved in https://github.com/vllm-project/vllm/pull/14155
